xtask/src/qemu.rs

```diff
@@ -5,18 +5,132 @@ use crate::pipe::Pipe;
 use crate::tpm::Swtpm;
 use crate::util::command_to_string;
 use crate::{net, platform};
-use anyhow::{anyhow, bail, Context, Result};
+use anyhow::{bail, Context, Result};
 use regex::bytes::Regex;
 use serde_json::{json, Value};
+use sha2::{Digest, Sha256};
 use std::env;
 use std::ffi::OsString;
-use std::io::{BufRead, BufReader, Read, Write};
+use std::io::{BufRead, BufReader, Cursor, Read, Write};
 use std::path::{Path, PathBuf};
 use std::process::{Child, Command, Stdio};
+use tar::Archive;
 use tempfile::TempDir;
+use ureq::Agent;
 #[cfg(target_os = "linux")]
 use {std::fs::Permissions, std::os::unix::fs::PermissionsExt};
 
+// Name of the ovmf-prebuilt release tag.
+const OVMF_PREBUILT_TAG: &str = "edk2-stable202211-r1";
+// SHA-256 hash of the release tarball.
+const OVMF_PREBUILT_HASH: &str = "b085cfe18fd674bf70a31af1dc3e991bcd25cb882981c6d3523d81260f1e0d12";
+
+/// Download `url` and return the raw data.
+fn download_url(url: &str) -> Result<Vec<u8>> {
+    let agent: Agent = ureq::AgentBuilder::new()
+        .user_agent("uefi-rs-ovmf-downloader")
+        .build();
+
+    // Limit the size of the download.
+    let max_size_in_bytes = 4 * 1024 * 1024;
+
+    // Download the file.
+    println!("downloading {url}");
+    let resp = agent.get(url).call()?;
+    let mut data = Vec::with_capacity(max_size_in_bytes);
+    resp.into_reader()
+        .take(max_size_in_bytes.try_into().unwrap())
+        .read_to_end(&mut data)?;
+    println!("received {} bytes", data.len());
+
+    Ok(data)
+}
+
+// Extract the tarball's files into `prebuilt_dir`.
+//
+// `tarball_data` is raw decompressed tar data.
+fn extract_prebuilt(tarball_data: &[u8], prebuilt_dir: &Path) -> Result<()> {
+    let cursor = Cursor::new(tarball_data);
+    let mut archive = Archive::new(cursor);
+
+    // Extract each file entry.
+    for entry in archive.entries()? {
+        let mut entry = entry?;
+
+        // Skip directories.
+        if entry.size() == 0 {
+            continue;
+        }
+
+        let path = entry.path()?;
+        // Strip the leading directory, which is the release name.
+        let path: PathBuf = path.components().skip(1).collect();
+
+        let dir = path.parent().unwrap();
+        let dst_dir = prebuilt_dir.join(dir);
+        let dst_path = prebuilt_dir.join(path);
+        println!("unpacking to {}", dst_path.display());
+        fs_err::create_dir_all(dst_dir)?;
+        entry.unpack(dst_path)?;
+    }
+
+    Ok(())
+}
+
+/// Update the local copy of the prebuilt OVMF files. Does nothing if the local
+/// copy is already up to date.
+fn update_prebuilt() -> Result<PathBuf> {
+    let prebuilt_dir = Path::new("target/ovmf");
```

Comment:
It was a little hidden where the files land and it took me more than just a quick glance to find out. I'd appreciate if you can introduce a new `OUTPUT_DIR` constant or so at the top of the file.

Reply:
done